[operator/java] Add instrumentation doc for java (#2253)
open-telemetry/opentelemetry.io: content/en/docs/k8s-operator/automatic.md

### Java

> The following command will create a basic Instrumentation resource that is

Reviewer: ```suggestion
The following command creates a basic Instrumentation resource that is
```

configured for instrumenting Java services.

```bash
kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: demo-instrumentation
spec:
  exporter:
    endpoint: http://demo-collector:4317
  propagators:
    - tracecontext
    - baggage
  sampler:
    type: parentbased_traceidratio
    argument: "1"
EOF
```

By default, the Instrumentation resource that auto-instruments Java services
uses `otlp` with the `grpc` protocol. This means that the configured endpoint
must be able to receive OTLP over `grpc`. Therefore, the example uses

> `http://demo-collector:4317`, which will connect to the `grpc` port of the

Reviewer: ```suggestion
`http://demo-collector:4317`, which connects to the `grpc` port of the
```

otlpreceiver of the Collector created in the previous step.

By default, the Java auto-instrumentation ships with

> [many many instrumentation libraries](/docs/instrumentation/java/automatic/#supported-libraries-frameworks-application-services-and-jvms).

Reviewer: ```suggestion
[many instrumentation libraries](/docs/instrumentation/java/automatic/#supported-libraries-frameworks-application-services-and-jvms).
```

> This makes instrumentation easy, but can result in too much or unwanted data. If

Reviewer: ```suggestion
This makes instrumentation easy, but could result in too much or unwanted data. If
```

there are any libraries you do not want to use you can set the
`OTEL_INSTRUMENTATION_[NAME]_ENABLED=false` where `[NAME]` is the name of the

> libary. If you know exactly which libraries you want to use you can disbale the

Reviewer: ```suggestion
library. If you know exactly which libraries you want to use, you can disable the
```

default libraries by setting `OTEL_INSTRUMENTATION_COMMON_DEFAULT_ENABLED=false`
and then use `OTEL_INSTRUMENTATION_[NAME]_ENABLED=true` where `[NAME]` is the
name of the library. For more details, see
[Suppressing specific auto-instrumentation](/docs/instrumentation/java/automatic/agent-config/#suppressing-specific-auto-instrumentation).

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: demo-instrumentation
spec:
  exporter:
    endpoint: http://demo-collector:4317
  propagators:
    - tracecontext
    - baggage
  sampler:
    type: parentbased_traceidratio

>     argument: "1"

Reviewer 2: Can you link to the documentation for these configuration options? I'm not familiar with where this is defined or how it translates to the environment variable based config scheme the otel java agent used.

Author: For propagators the valid values are `tracecontext;baggage;b3;b3multi;jaeger;xray;ottrace;none` (https://github.com/open-telemetry/opentelemetry-operator/blob/1a12ac7bb23f78a39a1b6a7f216faa034342b2f7/apis/v1alpha1/propagators.go#L19).  The operator will add them to the `OTEL_PROPAGATORS` environment variable.

For sampler the valid values are `always_on, always_off, traceidratio, parentbased_always_on, parentbased_always_off, parentbased_traceidratio, jaeger_remote, xray` (always_on, always_off, traceidratio, parentbased_always_on, parentbased_always_off, parentbased_traceidratio, jaeger_remote, xray).  The operator will add the sampler and its args to `OTEL_TRACES_SAMPLER` and `OTEL_TRACES_SAMPLER_ARG`.

I agree that this page doesn't discuss those aspects clearly, and neither does the Operator docs.  I will get an operator docs PR created to address this so that I have something to link to from this doc.

Reviewer: Is this blob required, or is this sorta a best practice?

Author: Are you asking if a propagator and sampler are required or only a best practice? They are best practices.

Reviewer: @jack-berg since @TylerHelmuth is adding this content, are you satisfied with the changes for now? We can file a tracking issue to link to the config options (or pull them in here) separately.

### Java

> The following command will create a basic Instrumentation resource that is

Reviewer 3: ```suggestion
The following command will create a basic instrumentation resource that is
```

Author: `Instrumentation` is a proper noun here I think since it is the type of the customer k8s resource we are creating.

Reviewer 3: Might it make sense for it to be in code font then?

Reviewer 3: Or possibly, on the first occurrence, link to https://github.com/open-telemetry/opentelemetry-operator/blob/main/docs/api.md#instrumentation?

Author: @chalin the page links to the Instrumentation CRDs with its first occurrence here: https://github.com/open-telemetry/opentelemetry.io/blob/92e34768e7885423b0fd719dcbb77a8b08f53573/content/en/docs/k8s-operator/automatic.md?plain=1#L103.  Is that acceptable?

Reviewer 3: Yes! Sorry, missed that.

configured for instrumenting Java services.

```bash
kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: demo-instrumentation
spec:
  exporter:
    endpoint: http://demo-collector:4317
  propagators:
    - tracecontext
    - baggage
  sampler:
    type: parentbased_traceidratio
    argument: "1"
EOF
```

> By default, the Instrumentation resource that auto-instruments Java services

Reviewer 3: ```suggestion
By default, the instrumentation resource that auto-instruments Java services
```

Author: Same here

uses `otlp` with the `grpc` protocol. This means that the configured endpoint
must be able to receive OTLP over `grpc`. Therefore, the example uses

> `http://demo-collector:4317`, which will connect to the `grpc` port of the

Reviewer 3: Make the link clickable / more-easily copiable?
```suggestion
<http://demo-collector:4317>, which will connect to the `grpc` port of the
```

Author: I don't think we want to make this clickable as it won't go anywhere useful.  Within the context of this page that address is only available from within your k8s cluster.

otlpreceiver of the Collector created in the previous step.

> By default the Java auto-instrumentation ships with

Reviewer 3: ```suggestion
By default, the Java auto-instrumentation ships with
```

[many many instrumentation libraries](/docs/instrumentation/java/automatic/#supported-libraries-frameworks-application-services-and-jvms).
This makes instrumentation easy, but can result in too much or unwanted data. If
there are any libraries you do not want to use you can set the
`OTEL_INSTRUMENTATION_[NAME]_ENABLED=false` where `[NAME]` is the name of the
libary. If you know exactly which libraries you want to use you can disbale the
default libraries by setting `OTEL_INSTRUMENTATION_COMMON_DEFAULT_ENABLED=false`
and then use `OTEL_INSTRUMENTATION_[NAME]_ENABLED=true` where `[NAME]` is the

> name of the library. See
> [Suppressing specific auto-instrumentation](/docs/instrumentation/java/automatic/agent-config/#suppressing-specific-auto-instrumentation)
> in the Java docs for more details

Reviewer 3: ```suggestion
name of the library. For more details, see
[Suppressing specific auto-instrumentation](/docs/instrumentation/java/automatic/agent-config/#suppressing-specific-auto-instrumentation).
```

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: demo-instrumentation
spec:
  exporter:
    endpoint: http://demo-collector:4317
  propagators:
    - tracecontext
    - baggage
  sampler:
    type: parentbased_traceidratio
    argument: "1"
  java:
    env:
      - name: OTEL_INSTRUMENTATION_KAFKA_ENABLED
        value: false
      - name: OTEL_INSTRUMENTATION_REDISCALA_ENABLED
        value: false
```

> [See the Java Agent Configuration docs for more details.](/docs/instrumentation/java/automatic/agent-config/)

Reviewer 3: ```suggestion
For more details, see [Agent Configuration](/docs/instrumentation/java/automatic/agent-config/).
```
